File: src/Network.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

public final class Network {
    private final Map<String, Node> nodes;
    private final Map<Node, Map<Node, Integer>> distances = new HashMap<>();

    public Network(String path) throws FileNotFoundException {
        nodes = new HashMap<>();

        generateNetwork(path);
    }

    private void generateNetwork(String path) throws FileNotFoundException {
        List<String[]> links = readCSV(path);

        for(String link[]:links) {
            Node a = generateNode(link[0]);
            Node b = generateNode(link[1]);
            a.addNeighbour(b);
            b.addNeighbour(a);
            Integer distance = Integer.parseInt(link[2]);

            if(distances.get(a) != null) { // entry already exists
                distances.get(a).put(b,distance);
            } else {
                Map<Node, Integer> entry = new HashMap<>();
                entry.put(b, distance);
                distances.put(a, entry);
            }
        }
    }

    /**
     * Search for Node with given name, return if found, else generate a new node and add it to the nodes list
     *
     * @param name Name for the node
     * @return  Node with given name
     */
    private Node generateNode(String name) {
        if (nodes.containsKey(name)) {
            return nodes.get(name);
        } else {
            Node node  = new Waypoint(name);
            nodes.put(name, node);
            return node;
        }
    }

    /**
     * Get a node from the nodes list with given name
     * @param name
     * @return node or null of not found
     */
    public Node getNodeByName(String name) {
        return nodes.get(name);
    }

    /**
     * Get distance between two nodes
     * @param from
     * @param to
     * @return integer distance between node from and node to
     */
    public int getDistance(Node from, Node to) {
        if(from == to) {
            return 0;
        }

        if(distances.containsKey(from) && distances.containsKey(to)) {
            if(distances.get(from).get(to) != null) {
                return distances.get(from).get(to);
            }

            if(distances.get(to).get(from) != null) {
                return distances.get(to).get(from);
            }
        }

        throw new NoSuchElementException();
    }

    public List<Node> getNodes() {
        return new ArrayList<Node>(nodes.values());
    }

    /**
     * Read CSV File with network information
     * File needs to be formattet like: "<Nodename A>, <Nodename B>, <int>"
     *
     * @return List with each entry beeing a size 3 array with the two nodenames and the distance between as entries
     */
    private List<String[]> readCSV(String path) throws FileNotFoundException {
        String csvSeparator = ",";

        BufferedReader br = new BufferedReader(new FileReader(path));

        try {
            String line;
            List<String[]> links = new ArrayList<>();
            while ((line = br.readLine()) != null) {
                String[] link = line.split(csvSeparator); // use comma as separator
                links.add(link);
            }
            return links;
        } catch (IOException e) {
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    private class Waypoint implements Node {
        private int currentDistance;
        private List<Node> neighbours;
        private String name;
        private boolean isVisited;
        private Node predecessor;

        public Waypoint(String name) {
            currentDistance = Integer.MAX_VALUE;
            neighbours = new LinkedList<>();
            this.name = name;
            isVisited = false;
            predecessor = null;
        }
        public Node getPredecessor() {
            return predecessor;
        }

        public void setPredecessor(Node predecessor) {
            this.predecessor = predecessor;
        }

        public int getCurrentDistance() {
            return currentDistance;
        }

        public void setCurrentDistance(int value) {
            currentDistance = value;
        }

        public List<Node> getNeighbours() {
            return neighbours;
        }

        public void addNeighbour(Node neighbour) {
            this.neighbours.add(neighbour);
        }

        public boolean isVisited() {
            return isVisited;
        }

        public void setAsVisited() {
            this.isVisited = true;
        }

        public void setAsUnvisited() {
            this.isVisited = false;
        }

        public String getName() {
            return name;
        }

        @Override
        public String toString() {
            return "Node{" +
                    "currentDistance=" + currentDistance +
                    ", name='" + name + '\'' +
                    ", isVisited=" + isVisited +
                    '}';
        }
    }
}

File: src/python/Network.py
import copy
import csv

from node import Node


class Network:
    __nodes = {}
    __distances = {}

    def __init__(self, path: str):
        self.__generate_network(path)

    def __generate_network(self, path, encoding='utf-8') -> None:
        with open(path, 'r', encoding=encoding) as csvfile:
            data = csv.reader(csvfile, delimiter=',')
            for row in data:
                a = self.__get_node_create_if_not_existing(row[0])
                b = self.__get_node_create_if_not_existing(row[1])

                a.add_neighbour(b)
                b.add_neighbour(a)

                distance = int(row[2])

                node_distance_dictionary = self.__distances.get(a)
                if node_distance_dictionary is not None:
                    node_distance_dictionary[b] = distance
                else:
                    entry = {b: distance}
                    self.__distances[a] = entry

    def __get_node_create_if_not_existing(self, name: str) -> Node:
        if name in self.__nodes:
            return self.__nodes[name]
        else:
            node = Node(name)
            self.__nodes[name] = node
            return node

    def get_node(self, name: str) -> Node:
        return self.__nodes[name]

    def get_distance(self, from_node: Node, to_node: Node) -> int:
        if from_node == to_node:
            return 0

        if from_node in self.__distances and to_node in self.__distances[from_node]:
            return self.__distances[from_node][to_node]

        if to_node in self.__distances and from_node in self.__distances[to_node]:
            return self.__distances[to_node][from_node]

    def get_nodes(self) -> dict:
        return copy.deepcopy(self.__nodes)
